runtime/drivers/clickhouse/olap.go

```diff
@@ -251,7 +251,70 @@ func (c *connection) InsertTableAsSelect(ctx context.Context, name, sql string,
 		})
 	}
 
-	// merge strategy is also not supported for clickhouse
+	if strategy == drivers.IncrementalStrategyMerge {
+		_, onCluster, err := informationSchema{c: c}.entityType(ctx, "", name)
+		if err != nil {
+			return err
+		}
+		onClusterClause := ""
+		if onCluster {
+			onClusterClause = "ON CLUSTER " + safeSQLName(c.config.Cluster)
+		}
+		// Get the engine info of the given table
+		engine, err := c.getTableEngine(ctx, name)
+		if err != nil {
+			return err
+		}
+		// Distributed table cannot be altered directly, so we need to alter the local table
+		if engine == "Distributed" {
+			name = localTableName(name)
+		}
+		// create temp table with the same schema
+		tempName := tempName(name)
+		err = c.Exec(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("CREATE OR REPLACE TABLE %s %s AS %s", safeSQLName(tempName), onClusterClause, name),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		// clean up the temp table
+		defer func() {
+			err = c.Exec(ctx, &drivers.Statement{
+				Query:    fmt.Sprintf("DROP TABLE %s %s", safeSQLName(tempName), onClusterClause),
+				Priority: 1,
+			})
+			if err != nil {
+				c.logger.Warn("clickhouse: failed to drop temp table", zap.String("name", tempName), zap.Error(err))
+			}
+		}()
+		// insert into temp table
+		err = c.Exec(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("INSERT INTO %s %s", safeSQLName(tempName), sql),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		// list partitions from the temp table
+		partitions, err := c.getTablePartitions(ctx, tempName)
+		if err != nil {
+			return err
+		}
+		// iterate over partitions and replace them in the main table
+		for _, part := range partitions {
+			// alter the main table to replace the partition
+			err = c.Exec(ctx, &drivers.Statement{
+				Query:    fmt.Sprintf("ALTER TABLE %s %s REPLACE PARTITION %s FROM %s", safeSQLName(name), onClusterClause, part, safeSQLName(tempName)),
```

Comment:
So this is more like a "partition overwrite" strategy than a merge? (I.e. it replaces partitions instead of upserting/merging rows). If yes, would it make sense to add a new `drivers.IncrementalStrategyPartitionOverwrite` instead of using `drivers.IncrementalStrategyMerge`?

In ClickHouse, am I correct in saying that a normal "merge" strategy (that does upserts, not partition replaces like this does) can be achieved by using the [`ReplacingMergeTree`](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree) table engine, and doesn't require special insert implementation?

Reply:
`ReplacingMergeTree` is eventually consistent so duplicate rows may exist temporarily until the background merges occur.
Current implementation of the merge strategy is pretty much similar to the DuckDB's but relies on a partition key instead of a custom one.

Reply:
> is eventually consistent so duplicate rows may exist temporarily until the background merges occur.
Got it, makes sense (though it sounds like an annoying property)

> but relies on a partition key instead of a custom one

So my question is if this means that it will replace/overwrite full partitions? I.e. can this lead to rows being removed (which is not possible in the DuckDB implementation). If it can lead to rows being removed, then I think we should call the strategy `partition_overwrite` instead (or something like that).

For reference, take a look at how DBT distinguishes between their `merge` and `insert_overwrite` incremental strategies: https://docs.getdbt.com/reference/resource-configs/bigquery-configs#merge-behavior-incremental-models

Reply:
> Got it, makes sense (though it sounds like an annoying property)

The background merge might be [triggered manually](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree)
> Although you can run an unscheduled merge using the `OPTIMIZE` query, do not count on using it, because the `OPTIMIZE` query will read and write a large amount of data

We can implement such approach for small datasets in ClickHouse `merge` strategy.

> I think we should call the strategy `partition_overwrite` instead

Agree, lets call it so or `insert_overwrite` if we want to follow some common syntax

Reply:
> We can implement such approach for small datasets in ClickHouse merge strategy.

It sounds like there might still be a small period of inconsistency even if we trigger it manually? Might be best to avoid the manual triggers.

I wonder if CH has a way to not serve new rows for a `ReplacingMergeTree` until the background replace has completed. Or some kind of watermark indicating up to what row ID that the table is consistent, which we could pass in metrics queries.

> Agree, lets call it so or `insert_overwrite` if we want to follow some common syntax

IMO `partition_overwrite` sounds more correct than `insert_overwrite` (which sounds like a convoluted way of saying "upsert"). But if you prefer to align with DBT terminology, that makes sense and also sounds good to me – you decide :)

Reply:
> small period of inconsistency even if we trigger it manually? 

the period depends on data volume, it might be significant if the volume is large.

> which we could pass in metrics queries

yes, `FINAL` keyword might be [passed in queries](https://clickhouse.com/docs/en/guides/replacing-merge-tree#querying-replacingmergetree) as a guarantee of correctness.

> To obtain correct answers, users will need to complement background merges with query time deduplication and deletion removal. This can be achieved using the FINAL operator.

But `FINAL` has a [performance overhead](https://clickhouse.com/docs/en/guides/replacing-merge-tree#final-performance) and needs to be used wisely